src/py21cmfast/drivers/coeval.py

```diff
@@ -0,0 +1,901 @@
+        true, as if the next box to be calculate has different shape, errors will occur
+        if memory is not cleaned. Note that internally, this is set to False until the
+        last iteration.
+    \*\*global_kwargs :
+        Any attributes for :class:`~py21cmfast.inputs.GlobalParams`. This will
+        *temporarily* set global attributes for the duration of the function. Note that
+        arguments will be treated as case-insensitive.
+
+    Returns
+    -------
+    coevals : :class:`~py21cmfast.outputs.Coeval`
+        The full data for the Coeval class, with init boxes, perturbed fields, ionized boxes,
+        brightness temperature, and potential data from the conservation of photons. If a
+        single redshift was specified, it will return such a class. If multiple redshifts
+        were passed, it will return a list of such classes.
+
+    Other Parameters
+    ----------------
+    regenerate, write, direc, random_seed :
+        See docs of :func:`initial_conditions` for more information.
+    """
+    if redshift is None and perturbed_field is None:
+        raise ValueError("Either redshift or perturb must be given")
+
+    direc, regenerate, hooks = _get_config_options(direc, regenerate, write, hooks)
+
+    singleton = False
+    # Ensure perturb is a list of boxes, not just one.
+    if perturbed_field is None:
+        perturbed_field = []
+    elif not hasattr(perturbed_field, "__len__"):
+        perturbed_field = [perturbed_field]
+        singleton = True
+
+    random_seed = (
+        initial_conditions.random_seed
+        if initial_conditions is not None
+        else random_seed
+    )
+
+    inputs = InputParameters(
+        user_params=user_params,
+        cosmo_params=cosmo_params,
+        astro_params=astro_params,
+        flag_options=flag_options,
+    )
+
+    iokw = {"regenerate": regenerate, "hooks": hooks, "direc": direc}
+
+    if initial_conditions is None:
+        initial_conditions = sf.initial_conditions(
+            user_params=inputs.user_params,
+            cosmo_params=inputs.cosmo_params,
+            random_seed=random_seed,
+            **iokw,
+        )
+
+    # We can go ahead and purge some of the stuff in the init_box, but only if
+    # it is cached -- otherwise we could be losing information.
+    with contextlib.suppress(OSError):
+        initial_conditions.prepare_for_perturb(
+            flag_options=flag_options, force=always_purge
+        )
+    if perturbed_field:
+        if redshift is not None and any(
+            p.redshift != z for p, z in zip(perturbed_field, redshift)
+        ):
+            raise ValueError("Input redshifts do not match perturb field redshifts")
+        else:
+            redshift = [p.redshift for p in perturbed_field]
+
+    kw = {
+        **{
+            "astro_params": inputs.astro_params,
+            "flag_options": inputs.flag_options,
+            "init_boxes": initial_conditions,
+        },
+        **iokw,
```

Comment:
Yes -- because every other low-level function other than `initial_conditions` requires passing `initial_conditions` instead of the `user` and `cosmo` params (because they now _require_ the initial conditions to be passed, which already has those params defined in it, so it's easier to take them from there, rather than having two ways to get the params and having to make sure they match.